node-metrics/src/service/data_state/mod.rs

```diff
@@ -210,6 +286,81 @@ pub fn create_block_detail_from_block(block: &BlockQueryData<SeqTypes>) -> Block
     }
 }
 
+// epoch_number_helper is a utility function that will determine the
+// epoch number based on the given [block_height], [epoch_starting_block],
+// and [num_blocks_per_epoch].
+//
+// This function will ensure that the epoch number is zero if the given
+// [block_height] is less than the [epoch_starting_block]. Otherwise it
+// will return the [epoch_from_block_number] result.
+fn epoch_number_helper(
+    block_height: u64,
+    epoch_starting_block: u64,
+    num_blocks_per_epoch: u64,
+) -> u64 {
+    if block_height < epoch_starting_block {
+        return 0;
+    }
+
+    epoch_from_block_number(block_height, num_blocks_per_epoch)
+}
+
+async fn perform_stake_table_epoch_check_and_update(
+    data_state: Arc<RwLock<DataState>>,
+    client: surf_disco::Client<hotshot_query_service::Error, Version01>,
+    hotshot_config: &PublicHotShotConfig,
+    block_height: u64,
+) -> Result<(), ProcessLeafError> {
+    // Are we in a new epoch?
+    // Do we need to replace our stake table?
+    tracing::debug!("processing block height: {}", block_height);
+    if let (Some(epoch_starting_block), Some(num_blocks_per_epoch)) = (
+        hotshot_config.epoch_start_block,
+        hotshot_config.epoch_height,
+    ) {
+        let previous_epoch = epoch_number_helper(
+            block_height.saturating_sub(1),
+            epoch_starting_block,
+            num_blocks_per_epoch,
+        );
+        let upcoming_epoch =
+            epoch_number_helper(block_height, epoch_starting_block, num_blocks_per_epoch);
+
+        if upcoming_epoch != previous_epoch {
+            tracing::debug!(
```

Comment:
If we straight up replace the stake table as soon as we enter a new epoch, there is going to be a period where we have blocks in our "last 50" window from both the old and new epoch, but we are interpreting the QC bitmaps using only the new stake table. So the participation numbers might go a little wonky the first 50 blocks of each epoch.

This is honestly not that big of a deal I think especially because we don't expect the stake table to change that much from one epoch to the next usually. But we should fix it eventually

Reply:
For clarity, the votes are transferred to users independently of the Block data transferred.  The Votes are transferred in a `BitVec` whose ordering is the same as the collection of our stored Node Identities.